ssp

```diff
@@ -8363,29 +8372,68 @@ sub get_hostname {
     sub json {
         return $json if $json;
         return if $tried_import;
-        $json = import_json();
+        $json = load_module_with_fallbacks(
+            'needed_subs' => [qw{encode decode}],
+            'modules'     => [qw{Cpanel::JSON::XS JSON::XS JSON::PP}],
+        );
         return $json;
     }
+}
 
-    sub import_json {
-        return if $tried_import;
-        $tried_import = 1;
-        my $try = _choose_json_package();
-        eval {
-            return unless $try->can('decode');
-            return unless $try->can('encode');
-            1;
-        } and return $try;
+# SUB load_module_with_fallbacks(
+#   'modules'     => [ 'module1', 'module2', ... ],
+#   'needed_subs' => [ 'do_needful', ... ],
+#   'fallback'    => sub { *do_needful = sub { ... }; return; },
+#   'fail_safe'   => 1,
+# );
+#
+# Input is HASH of options:
+#   'modules'     => ARRAYREF of SCALAR strings corresponding to module names to attempt to import. These are attempted first.
+#   'needed_subs' => ARRAYREF of SCALAR strings corresponding to subroutine names you need defined from the module(s).
+#   'fallback'    => CODEREF which defines the needed subs manually. Only used if all modules passed in above fail to load. Optional.
+#   'fail_safe'   => BOOL whether or not to die if you fail to load the needed subs/modules via all available methods.
+#
+# Returns the module/namespace that loaded correctly, throws if all available attempts at finding the desired needed_subs subs fail (unless fail_safe passed).
+sub load_module_with_fallbacks {
+    my %opts = @_;
+    my $namespace_loaded;
+    foreach my $module2try ( @{ $opts{'modules'} } ) {
+
+        # Don't 'require' it if we already have it.
+        my $inc_entry = join( "/", split( "::", $module2try ) ) . ".pm";
+        if ( !$INC{$module2try} ) {
+            local $@;
+            next if !eval "require $module2try; 1";
+        }
+
+        # Check if the imported modules 'can' do the job
+        next if ( scalar( grep { $module2try->can($_) } @{ $opts{'needed_subs'} } ) != scalar( @{ $opts{'needed_subs'} } ) );
+
+        # Ok, we're good to go!
+        $namespace_loaded = $module2try;
+        last;
     }
 
-    sub _choose_json_package {
-        return 'Cpanel::JSON::XS' if $INC{'Cpanel/JSON/XS.pm'};
-        return 'JSON::XS'         if $INC{'JSON/XS.pm'};
-        return 'Cpanel::JSON::XS' if eval { require Cpanel::JSON::XS; 1; };
-        return 'JSON::XS'         if eval { require JSON::XS; 1; };
-        return 'JSON::PP'         if eval { require JSON::PP; 1; };
-    }
+    # Fallback to coderef, but don't do sanity checking on this, as it is presumed the caller "knows what they are doing" if passing a coderef.
+    if ( !$namespace_loaded ) {
+        if ( !$opts{'fallback'} || ref $opts{'fallback'} != 'CODE' ) {
```

Comment:
The multi-line warning is a bit too verbose for SSP, and it currently doesn't directly connect the failure to load the module as the reason that a check can't be performed:
```
[WARN] * SSP::load_module_with_fallbacks():
        Failed to load any of the following modules:
        YAML::Syck
        OR they did not have the wanted subroutines defined in the modules:
        LoadFile
        ...and no fallback subroutines exist within SSP to accomodate this.
[WARN] * SSP can't check rpm.versions.d/local.versions for customization -- Try using /usr/local/cpanel/3rdparty/bin/perl
```

They're printed consecutively but consecutive lines in SSP aren't necessarily related to each other, so it's not obvious that the failure to load YAML::Syck is the reason that SSP can't check rpm.versions.d/local.versions.

I think the solution to that is to be able to provide some additional help text to load_module_with_fallbacks().  For example:
```
my $yaml_module = load_module_with_fallbacks(
        'needed_subs' => [qw{LoadFile}],
        'modules' => [qw{YAML::Syck}],
        'fail_warning' => 'SSP can\'t check rpm.versions.d/local.versions for customization -- Try using /usr/local/cpanel/3rdparty/bin/perl',
```
Also, IF there is no fail_warning, do not print any warning.  This would allow a silent failure, if necessary.

Example:
```
[WARN] * Missing Perl Module: YAML::Syck -- SSP can't check rpm.versions.d/local.versions for customization -- Try using /usr/local/cpanel/3rdparty/bin/perl
```